Allow None Type for safe_eval

markafarrell opened the pull request:
##### SUMMARY
In some scenarios safe_eval fails with the error `NameError("name 'None' is not defined")`. This causes valid JSON to not be correctly evaluated into a python dictionary.

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
safe_eval

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

```

markafarrell committed:
also allow None Type for safe eval

jimi-c commented:
@markafarrell do you have an example playbook that triggers this bug? We'd like to add some tests and we're also considering whether this is common enough of an occurrence to require backporting.

Thanks!

markafarrell commented:
@jimi-c 

> @markafarrell do you have an example playbook that triggers this bug? We'd like to add some tests and we're also considering whether this is common enough of an occurrence to require backporting.
> 
> Thanks!

Unfortunately I couldn't replicate the issue with a simple playbook. I have some complex Jinja2 templates that generate a JSON object and couldn't exactly track down what was causing the issue.

I don't believe this issue was present before b9b0b230150eceb442c34c917d9e852d5e8b7371 (i.e. adding `'__builtins__': {},` to `OUR_GLOBALS`) so backporting shouldn't be required

abadger commented:
B9b023 was backported to 2.8, 2.7, and 2.6 so this fix will have to be backported as well if that is the change that introduced the problem.  Are you able to verify either that b9b0b23 is there first commit to show the issue with your complex playbook or checkout the current stable-2.8 branch and reproduce the issue there?  I'd need to get this backported and merged today/tonight or skip the next release in order to get this in.

markafarrell commented:
@abadger I can confirm that everything was working from the commit before b9b0b23 (99f9f49)

Note: The reason for the failure in the playbook is caused by the generated data not being evaluated into a valid python object.

See below:

Using commit 99f9f49
```
(.venv) bash-4.2# git log
commit 99f9f49ecafe3634652282139b4659e855ad8373
Author: Paul Gear <github@libertysys.com.au>
Date:   Fri Jun 7 05:05:01 2019 +1000

    Correct grammar (#57385)
    
    "it's" == "it is"
    "its" == "belonging to it"
```
```
markfarrell@t000088258672w0:~/Projects/epg-manage-config$ make tests/test-generate-netconf-day-1.yml 
Virtual Environment already exists. Not creating.
Cloning into '.epg-model-inventory'...
remote: Counting objects: 4466, done.
remote: Compressing objects: 100% (4019/4019), done.
remote: Total 4466 (delta 2326), reused 271 (delta 148)
Receiving objects: 100% (4466/4466), 4.11 MiB | 0 bytes/s, done.
Resolving deltas: 100% (2326/2326), done.
Already on 'master'
Running Playbook: tests/test-generate-netconf-day-1.yml
 [ERROR]: [DEPRECATION WARNING]: The TRANSFORM_INVALID_GROUP_CHARS settings is set to allow bad characters in group names by default, this will change, but still be user
configurable  on deprecation. This feature will be removed in version 2.10. Deprecation warnings can be disabled by setting deprecation_warnings=False in ansible.cfg.
[WARNING]: Invalid characters were found in group names but not replaced, use -vvvv to see details

[DEPRECATION WARNING]: The TRANSFORM_INVALID_GROUP_CHARS settings is set to allow bad characters in group names by default, this will change, but still be user configurable 
on deprecation. This feature will be removed in version 2.10. Deprecation warnings can be disabled by setting deprecation_warnings=False in ansible.cfg.
 [WARNING]: Invalid characters were found in group names but not replaced, use -vvvv to see details


PLAY [epg_2x] *****************************************************************************************************************************************************************

TASK [Remove the temporary generated JSON file] *******************************************************************************************************************************
changed: [DEV_EPG_1]

TASK [Remove the temporary generated XML file] ********************************************************************************************************************************
changed: [DEV_EPG_1]

TASK [include the role] *******************************************************************************************************************************************************

TASK [/data/tests/../ : Fail if required variable not defined] ****************************************************************************************************************
skipping: [DEV_EPG_1]

TASK [/data/tests/../ : Fail if required variable not defined] ****************************************************************************************************************
skipping: [DEV_EPG_1]

TASK [/data/tests/../ : Expand YANG input template for day-1] *****************************************************************************************************************
ok: [DEV_EPG_1]

TASK [/data/tests/../ : Write YANG input to file.] ****************************************************************************************************************************
changed: [DEV_EPG_1]

TASK [/data/tests/../ : Create tempfile for interim JSON] *********************************************************************************************************************
changed: [DEV_EPG_1]

TASK [/data/tests/../ : Write out interim JSON] *******************************************************************************************************************************
changed: [DEV_EPG_1]

TASK [/data/tests/../ : Define YANG files and search path] ********************************************************************************************************************
ok: [DEV_EPG_1]

TASK [/data/tests/../ : Generate XML] *****************************************************************************************************************************************
ok: [DEV_EPG_1]

TASK [/data/tests/../ : Remove intermediate json] *****************************************************************************************************************************
changed: [DEV_EPG_1]

TASK [Write out the YANG input JSON] ******************************************************************************************************************************************
changed: [DEV_EPG_1]

TASK [Make XML config pretty] *************************************************************************************************************************************************
changed: [DEV_EPG_1]

TASK [Write out the Generated XML] ********************************************************************************************************************************************
changed: [DEV_EPG_1]

TASK [Compare file contents] **************************************************************************************************************************************************
changed: [DEV_EPG_1 -> localhost]

TASK [Check return code] ******************************************************************************************************************************************************
skipping: [DEV_EPG_1]

PLAY RECAP ********************************************************************************************************************************************************************
DEV_EPG_1                  : ok=13   changed=10   unreachable=0    failed=0    skipped=3    rescued=0    ignored=0   
```
JSON generated from template
```
{
    "ericsson-asp-ipr:asp": {
        "default_asp_attr": [
            {
                "default_asp_attr": "management/1",
                "shutdown": false
            },
            {
                "default_asp_attr": "1/1",
                "shutdown": false
            },
            {
                "default_asp_attr": "2/1",
                "shutdown": false
            },
            {
                "default_asp_attr": "3/1",
                "shutdown": false
            },
            ....
```

Using commit b9b0b23
```
(.venv) bash-4.2# git log
commit b9b0b230150eceb442c34c917d9e852d5e8b7371
Author: Brian Coca <bcoca@users.noreply.github.com>
Date:   Thu Jun 6 15:36:22 2019 -0400

    safe_eval fix (#57188)
    
    * just dont pass locals
    
     - also fix globals
     - added tests
    
    * fixed tests
```
```
markfarrell@t000088258672w0:~/Projects/epg-manage-config$ make tests/test-generate-netconf-day-1.yml 
Virtual Environment already exists. Not creating.
Cloning into '.epg-model-inventory'...
remote: Counting objects: 4466, done.
remote: Compressing objects: 100% (4019/4019), done.
remote: Total 4466 (delta 2325), reused 271 (delta 148)
Receiving objects: 100% (4466/4466), 4.11 MiB | 249.00 KiB/s, done.
Resolving deltas: 100% (2325/2325), done.
Already on 'master'
Running Playbook: tests/test-generate-netconf-day-1.yml
 [ERROR]: [DEPRECATION WARNING]: The TRANSFORM_INVALID_GROUP_CHARS settings is set to allow bad characters in group names by default, this will change, but still be user
configurable  on deprecation. This feature will be removed in version 2.10. Deprecation warnings can be disabled by setting deprecation_warnings=False in ansible.cfg.
[WARNING]: Invalid characters were found in group names but not replaced, use -vvvv to see details

[DEPRECATION WARNING]: The TRANSFORM_INVALID_GROUP_CHARS settings is set to allow bad characters in group names by default, this will change, but still be user configurable 
on deprecation. This feature will be removed in version 2.10. Deprecation warnings can be disabled by setting deprecation_warnings=False in ansible.cfg.
 [WARNING]: Invalid characters were found in group names but not replaced, use -vvvv to see details


PLAY [epg_2x] *****************************************************************************************************************************************************************

TASK [Remove the temporary generated JSON file] *******************************************************************************************************************************
changed: [DEV_EPG_1]

TASK [Remove the temporary generated XML file] ********************************************************************************************************************************
changed: [DEV_EPG_1]

TASK [include the role] *******************************************************************************************************************************************************

TASK [/data/tests/../ : Fail if required variable not defined] ****************************************************************************************************************
skipping: [DEV_EPG_1]

TASK [/data/tests/../ : Fail if required variable not defined] ****************************************************************************************************************
skipping: [DEV_EPG_1]

TASK [/data/tests/../ : Expand YANG input template for day-1] *****************************************************************************************************************
ok: [DEV_EPG_1]

TASK [/data/tests/../ : Write YANG input to file.] ****************************************************************************************************************************
changed: [DEV_EPG_1]

TASK [/data/tests/../ : Create tempfile for interim JSON] *********************************************************************************************************************
changed: [DEV_EPG_1]

TASK [/data/tests/../ : Write out interim JSON] *******************************************************************************************************************************
changed: [DEV_EPG_1]

TASK [/data/tests/../ : Define YANG files and search path] ********************************************************************************************************************
ok: [DEV_EPG_1]

TASK [/data/tests/../ : Generate XML] *****************************************************************************************************************************************
fatal: [DEV_EPG_1]: FAILED! => 
  msg: 'An unhandled exception occurred while running the lookup plugin ''yang_json2xml''. Error was a <type ''exceptions.TypeError''>, original message: string indices must be integers'

PLAY RECAP ********************************************************************************************************************************************************************
DEV_EPG_1                  : ok=7    changed=5    unreachable=0    failed=1    skipped=2    rescued=0    ignored=0   

Makefile:205: recipe for target 'tests/test-generate-netconf-day-1.yml' failed
make: *** [tests/test-generate-netconf-day-1.yml] Error 2
```
JSON generated from template
```
{'ericsson-contexts-ipos:contexts': {'ericsson-context-ipos:context': [{'description': u'local context', 'context-name': u'local', 'ip': {'ericsson-dns-ipos:domain-lookup': [None], 'ericsson-acl-ipos:access-list': [{'access-list': u'protectLocalIn', 'description': u'ACL Protection for local context', 'seq': [{'acl-protocol': u'tcp', 'dst-any': [N.....
```

bcoca commented:
FYI ` 'ip': {'ericsson-dns-ipos:domain-lookup': [None], ` <= is invalid JSON, as `None` is a Python type but JSON uses `null` so this was working because we reevaluate as Python AFTER the JSON is interpreted.

abadger commented:
@markafarrell Cool. We've decided to backport and slip release of 2.8.2, 2.7.12, and 2.6.18 by one week.

markafarrell commented:
@bcoca 
It's very strange

The Jinja2 template looks like the below:

```
  "ericsson-contexts-ipos:contexts": {
    "ericsson-context-ipos:context": [
      {
        "context-name": "local",
        "description": "local context",
        "ip": {
          "ericsson-dns-ipos:domain-lookup": [null]
...
```
somewhere during the templating process something strange is happening. (Also note double quotes changing to single quotes)
